File: Data Structures & Algorithms/combination-target-sum/submission-2.js
class Solution {
    nums = []
    res = []
    target = -1
    /**
     * @param {number[]} nums
     * @param {number} target
     * @returns {number[][]}
     */
    combinationSum(nums, target) {
        this.nums = nums
        this.target = target

        let state = []
        this.foo(0, state)

        return this.res
    }

    foo(sum, state) {
        if(sum === this.target) {
            this.res.push([...state])
            return
        }

        for(let i = 0; i < this.nums.length; i++) {
            let n = this.nums[i]
            console.log()
            state.push(n)
            this.foo(sum + n, state)
            state.pop()
            this.foo(sum, state)
        }
    }
}

File: Data Structures & Algorithms/two-integer-sum/submission-2.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        hash_map: dict[int, int] = {}

        i = 0
        while i < len(nums):
            n = nums[i]
            if target - n in hash_map:
                return [hash_map[target - n], i]
            hash_map[n] = i

            i += 1

        return []


